readme.txt
=== NextCellent Gallery - NextGEN Legacy ===
Contributors: WPGetReady
Tags:gallery,image,images,photo,photos,picture,pictures,slideshow,flash,media,thumbnails,photo-albums,NextGEN-gallery,NextGEN,nextcellent-gallery,nextcellent
Requires at least: 3.5
Tested up to: 3.8
Stable tag: trunk
License: GPLv2

== Description ==

= WordPress Gallery Plugin =

= V1.9.16 - 08.01.2014 =

 What's in it for you?

* Folder and Image Management improved (credits to Niko Strijbol)
* German translation (credits to Niko Strijbol)
* Improved style for WP 3.8 (credits to Niko Strijbol)
* Improper call to method as static when method belongs to instance.
* Code cleaning, proper call to static fixed

VERY IMPORTANT: Read ON!
-----------------------

**NextCellent Gallery provides backward compatibility for older NextGEN until version 1.9.13** . 

- **this plugin will gracefully deactivate if detects NextGEN is working (any version) to avoid compatibility issues**.

- Please remember to **READ THE FAQ!!!** <u>Issues for failing to read the FAQ will be IGNORED!!!</u>

- If you like it, please spread the word and rate it accordingly. I guess a lot of annoyed users can take advantage of NextCellent. Thank you!


= So, who is going to use this plugin? =

- The aim of NextCellent is to provide an alternative for traditional NextGEN users to keep their sites updated without breaking compatibility.  Older subplugins will be also compatible (those which were compatible with version 1.9.13 or earlier).

- Users who wants to stay working with older NextGEN version (1.9.13 or older) until they're confident enough to make the jump to newest 2.x.x version

- Users wishing avoid NextGEN 2.x.x and beyond for the moment.

- Site managers handling several sites using specialized software as InfiniteWP or ManageWP, wanting to keep their sites running with automatic updates.

- Users prefering an alternative development path with backward compatibility.

- Developers wanting to suggest or colaborate improving the code.


= What do you get with NextCellent Gallery? =

- Backward compatibility with NextGEN plugin version (1.9.13)

- Slow evolving code path, made only by NextGEN user's suggestions. Yep, you read it right: *slow* in counterpart as *fast*. Older code is good enough to keep a community and it worked (and works) for most people. Versions will rollup about once a month.

- A reliable way to work with already installed NextGEN galleries.

- A place for updating the plugin without using FTP manual updates, but Wordpress plugin repository.

- Alternative path preserving backward compatibility (while possible).

- **Development on Bitbucket open to developers suggestions**. (https://bitbucket.org/wpgetready/nextcellent). You are free to download , test and make suggestions and request.

Being said that, here are the usual classic features:

NextCellent  Gallery provides a powerful engine for uploading and managing galleries of images, with the ability to batch upload, import meta data, add/delete/rearrange/sort images, edit thumbnails, group galleries into albums, and more. It also provides two front-end display styles (slideshows and thumbnail galleries), both of which come with a wide array of options for controlling size, style, timing, transitions, controls, lightbox effects, and more.

= NextCellent WordPress Gallery Plugin Features =

*Upload Galleries*

* Our WordPress gallery plugin offers diverse and powerful functionality for getting images from your desktop to your website. You can easily upload batches of images via a standard WordPress-style uploader, or upload images via zip file or FTP. NextCellent will automatically import your images meta data.

*Manage Galleries*

* Centralized gallery management. Enjoy a single location where you can see and manage all your galleries.
* Edit galleries.  Add or exclude images, change gallery title and description, reorder of images, resize thumbnails.
* Thumbnail Management. Turn thumbnail cropping on and off, customize how individual thumbnails are cropped, and bulk resize thumbnails across one or more galleries.
* Edit Individual Images. Edit meta data and image tags, rotate images, and exclude images.
* Watermarks. Quickly add watermarks to batches or galleries of images.
* Albums. Create and organize collections of galleries, and display them in either compact or extended format.

*Display Galleries*

* Two Gallery Types. Choose between two main display styles: Slideshow and Thumbnail, and allow visitors to toggle between the two.
* Slideshow Galleries. Choose from a vast array of options for slideshows, including slideshow size, transition style, speed, image order, and optional navigation bar.
* Thumbnail Galleries. Choose from a wide range of options to customize thumbnail galleries, including 5 different lightboxes for individual images, optional thumbnail cropping and editing, thumbnail styles, captions, and more.
* Single Image Displays. Display and format single images.
* Work with Options Panel or Shortcodes.

= NextCellent WordPress Gallery Plugin Community & Extensions =

NextGEN has been the dominant WordPress gallery plugin for years. 
Being said that, NextCellent will provide backward compatibility for NextGEN 1.9.13 and it will evolve according user requirements.

As a result, there is large and great community of users and developers, as well as a large number of dedicated extension plugins. For a list of extension plugins, just search for NextGEN in the WordPress.org plugin repository, or visit our <a href="http://www.NextGEN-gallery.com/NextGEN-gallery-extension-plugins/">Complete List of NextGEN Extension Plugins</a>.

== Credits ==

Copyright:<br>
WpGetReady 2013<br>
Photocrati Media 2012<br>
Alex Rabe 2007-2011<br>

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program; if not, write to the Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

** Please note **

NextCellent Gallery's flash slideshow option is powered by the JW Image Rotator from Long Tail Video. The Image Rotator is NOT provided with this plugin. For more information, see the Long Tail Video website: http://www.longtailvideo.com.

== Installation ==

1. 	Download, and install from the usual Wordpress Panel.

2. 	Remember: plugin won't activate if NextGEN is installed and activated. A message will remind you this situation. Please deactivate any NextGEN plugin before installing NextCellent.

3.	From your Wordpress Dashboard, go to Gallery > Add Gallery/Images > Follow the on-screen cues.

3. 	Go to a post/page, and select the NextCellent Gallery button from the Kitchen Sink. Follow the on-screen cues to select, adjust, and publish your gallery.

That's it ... have fun! 

== Screenshots ==

1. Screenshot Admin Area
2. Screenshot Album Selection
3. Screenshot Shutter Effect
4. Screenshot Watermark function
5. Screenshot Flexible template layout
6. Screenshot Show Exif data

== Shortcode ==

NextCellent Plugin Shortcodes are 100% backward compatible with older NGG shortcodes.

= Examples =

*Use Image Tags to Create Galleries/Albums* - [ nggtags album=WordPress,Cologne,Ireland ]

*Display Captions in Thumbnail Galleries* - [ nggallery id=1 template=caption ]

*Basic Filmstrip Galleries* - [ nggallery id=2 template=carousel images=7 ]

*Display Exif Data* - [ imagebrowser id=28 template=exif ]

*Sort Images in a Gallery Based on Their Tags* - [ nggtags gallery=cologne,wordpress,.. ]

*Add Tag Clouds* - [ tagcloud]

*Single Pic Options* - [ singlepic id=x w=width h=height mode=web20|watermark float=left|right ]

*Template Engine for Gallery Types*<br>
[ nggallery id=1 template=sample1 ]<br>
[ nggallery id=1 template=sample2 ]<br>
[ nggallery id=1 template=sample3 ]<br>
[ nggallery id=1 template=sample4 ]<br>
[ nggallery id=1 template=sample5 ]<br>
[ nggallery id=1 template=sample6 ]<br>

*Integration with Third Party Plugins*<br>
[ monoslideshow id=1 w=450 h=350 ]<br>
[ nggallery id=1 template=galleryview images=0 ]<br>
[ media id=6 width=320 height=240 plugins=revolt-1 ]<br>
[ media id=3 width=320 height=240 plugins=rateit-2 ]<br>

== Frequently Asked Questions ==

= What are the SIGNIFICATIVE differences between NextCellent and NextGEN 1.9.13? =

- Plugin will detect if NextGEN is installed and it will deativate itself. You'll need to deactivate NextGEN in order to properly use NextCellent.
- Plugin will deactivate if you install and activate any NextGEN version (any version).
- Minimal text changes to let the user know what plugin is running. This would affect translated texts in some cases. 

= What are the SIGNIFICATIVE differences between this NextCellent and NextGEN 2.x.x? =

- New NextGEN (2.0.x and ahead) uses a completely redesigned framework. Compatibility at code level could be close to none. Being said that, most/all plugins depending on NextGEN would need an internal redesign to work with NextGEN 2.0.x

= I'm going to install the plugin on a fresh site. Any recommendations? = 

- None than usual. Follow the installation procedure!

= I'm currently using NextGEN. How can I switch from NextGEN to NextCellent? =

- Install NextCellent plugin but do not activate. If you accidentally activate it will tell you that it cannot be activated along with NextGEN.
- Deactivate NextGEN plugin.
- Activate NextCellent
- Play around

= Can I go back and switch to NextGEN if NextCellent is not convincing? =

- Absolutely, just activate NextGEN and NextCellent will deactivate itself automatically

= I prefer using NextGEN, how can I uninstall NextCellent plugin without loosing my pictures/album/galleries? =

- Assuming you have both plugins installed:
- If you are unsure, backup the site first(!)
- Be sure NextCellent is deactivated
- MANUALLY remove NextCellent plugin folder (usually nextcellent-gallery-nextgen-legacy). **DO NOT** use traditional Wordpress uninstall procedure. Remember: if you run the uninstall procedure you are wipping out your images/galleries(!)
- Reactivate NextGEN (if it was deactivated)
- That's all

= I switched to NextCellent, how can I uninstall the NextGEN plugin without loosing my pictures/album/galleries? =

- Assuming you have both plugins installed:
- If you are unsure, I backup the site first(!)
- Be sure NextGEN is deactivated
- MANUALLY remove NextGEN plugin folder(usually NextGEN-gallery) . **DO NOT** use traditional Wordpress uninstall procedure. Remember: if you run the uninstall procedure you are wipping out your images/galleries(!)
- Activate NextCellent (if it was deactivated)
- That's all

= I have NextGEN 1.9.13 or older, should I install NextCellent? =

- Good question. NextCellent makes no significatives differences at this current stage, except minimal text changes and NextGEN detection to avoid user confusion. Aside that, it should be identical, handling your galleries without problems. This is a starting point to anyone using older NextGEN version.

= If I have NextGEN 2.x.x, should I install NextCellent? =

- In case NextGEN is working right for you and you had never a problem into transition from 1.9.x to 2.x.x, you can safely ignore NextCellent.
- If you are coming from older NextGEN versions (1.9.13 and earlier) and for some reason you want an alternative development path, give this plugin a try.
- There is no guarantee about compatibility with future NextGEN releases.
 
= I am a site owner , would NextCellent be a good option? =

- If you want to keep the older plugin version, choose NextCellent.
- If you have a lot of sites with customized galleries, choose NextCellent. 
- If you have a site with a lot of customized code connected with older NextGEN versions, choose NextCellent.
- If you are getting performance problems, go with NextCellent.
- If you need peace of mind right now, go with NextCellent.
- If you are managing large amounts of sites, and you need update-and-forget without FTP (like InfiniteWP or ManageWP), go with NextCellent.
- NextCellent is a place where you can download and install currently an older, stable version. And when I say 'older' is completely different of saying obsolete. We (the users an me) are taking a different development path, picking up the old, good code and working from there.  

= I am a developer , would be wise to create plugins based on NextCellent? =

- That is another good question. Since NextCellent is a branch from NextGEN 1.9.13, I will do my best to keep it stable. I can assure the code will lack of dramatic changes over the time, unless there is a big need to do that.

= Are you related with Alex Rabe or PhotoCrati? =


changelog.txt
NextCellent Gallery
by WPReady

= V1.9.14 - 01.09.2013 =
* NEW:     Deactivate if NextGen is installed and activated


nggallery.php
<?php
/*
Plugin Name: NextCellent Gallery
Plugin URI: http://www.wpgetready.com/nextcellent-gallery
Description: A Photo Gallery for WordPress providing NextGEN legacy compatibility from version 1.9.13
Author: WPGReady based on Alex Rabe & PhotoCrati work.
Author URI: http://www.wpgetready.com
Version: 1.9.16

Copyright (c) 2007-2011 by Alex Rabe & NextGEN DEV-Team
Copyright (c) 2012 Photocrati Media
Copyright (c) 2013 WPGetReady

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

// Stop direct call
if(preg_match('#' . basename(__FILE__) . '#', $_SERVER['PHP_SELF'])) { die('You are not allowed to call this page directly.'); }

/**
 * Indicates that a clean exit occured. Handled by set_exception_handler
 */
if (!class_exists('E_Clean_Exit')) {
	class E_Clean_Exit extends RuntimeException
	{

	}
}

//If NextGEN is activated, deactivate this plugin, and warn about it!
check_nextgen::nextgen_activated();

/**
 * Loads the NextGEN plugin
 */
if (!class_exists('nggLoader')) {
	class nggLoader {

		var $version     = '1.9.16';
		var $dbversion   = '1.8.1';
		var $minimum_WP  = '3.5';
        //TODO: to be removed next iterations, since Photocrati disabled
		var $donators    = 'http://www.nextgen-gallery.com/donators.php';
		var $options     = '';
		var $manage_page;
		var $add_PHP5_notice = false;

		function nggLoader() {

			// Stop the plugin if we missed the requirements
			if ( ( !$this->required_version() ) || ( !$this->check_memory_limit() )  )
				return;

			// Set error handler
			set_exception_handler(array(&$this, 'exception_handler'));

			// Get some constants first
			$this->load_options();
			$this->define_constant();
			$this->define_tables();
			$this->load_dependencies();
			$this->start_rewrite_module();

			$this->plugin_name = basename(dirname(__FILE__)).'/'.basename(__FILE__);

			// Init options & tables during activation & deregister init option
			register_activation_hook( $this->plugin_name, array(&$this, 'activate') );
			register_deactivation_hook( $this->plugin_name, array(&$this, 'deactivate') );

			// Register a uninstall hook to remove all tables & option automatic
			register_uninstall_hook( $this->plugin_name, array(__CLASS__, 'uninstall') );

			// Start this plugin once all other plugins are fully loaded
			add_action( 'plugins_loaded', array(&$this, 'start_plugin') );

			// Register_taxonomy must be used during the init
			add_action( 'init', array(&$this, 'register_taxonomy') );
			add_action( 'wpmu_new_blog', array(&$this, 'multisite_new_blog'), 10, 6);

			// Add a message for PHP4 Users, can disable the update message later on
			if (version_compare(PHP_VERSION, '5.0.0', '<'))
				add_filter('transient_update_plugins', array(&$this, 'disable_upgrade'));

			//Add some links on the plugin page
			add_filter('plugin_row_meta', array(&$this, 'add_plugin_links'), 10, 2);

			// Check for the header / footer
			add_action( 'init', array(&$this, 'test_head_footer_init' ) );

			// Show NextGEN version in header
			add_action('wp_head', array('nggGallery', 'nextgen_version') );

			// Handle upload requests
			add_action('init', array(&$this, 'handle_upload_request'));
		}

		function start_plugin() {

			global $nggRewrite;

			// Load the language file
			$this->load_textdomain();

			// All credits to the tranlator
			$this->translator  = '<p class="hint">'. __('<strong>Translation by : </strong><a target="_blank" href="http://alexrabe.de/wordpress-plugins/nextgen-gallery/languages/">See here</a>', 'nggallery') . '</p>';
			$this->translator .= '<p class="hint">'. __('<strong>This translation is not yet updated for Version 1.9.0</strong>. If you would like to help with translation, download the current po from the plugin folder and read <a href="http://alexrabe.de/wordpress-plugins/wordtube/translation-of-plugins/">here</a> how you can translate the plugin.', 'nggallery') . '</p>';

			// Content Filters
			add_filter('ngg_gallery_name', 'sanitize_title');

			// Check if we are in the admin area
			if ( is_admin() ) {

				// Pass the init check or show a message
				if (get_option( 'ngg_init_check' ) != false )
					add_action( 'admin_notices', create_function('', 'echo \'<div id="message" class="error"><p><strong>' . get_option( "ngg_init_check" ) . '</strong></p></div>\';') );

			} else {

				// Add MRSS to wp_head
				if ( $this->options['useMediaRSS'] )
					add_action('wp_head', array('nggMediaRss', 'add_mrss_alternate_link'));

				// Look for XML request, before page is render
				add_action('parse_request',  array(&$this, 'check_request') );

				// Add the script and style files
				add_action('wp_enqueue_scripts', array(&$this, 'load_scripts') );
				add_action('wp_enqueue_scripts', array(&$this, 'load_styles') );

			}
		}

		function check_request( $wp ) {

			if ( !array_key_exists('callback', $wp->query_vars) )
				return;

			if ( $wp->query_vars['callback'] == 'imagerotator') {
				require_once (dirname (__FILE__) . '/xml/imagerotator.php');
				exit();
			}

			if ( $wp->query_vars['callback'] == 'json') {
				require_once (dirname (__FILE__) . '/xml/json.php');
				exit();
			}

			if ( $wp->query_vars['callback'] == 'image') {
				require_once (dirname (__FILE__) . '/nggshow.php');
				exit();
			}

			//TODO:see trac #12400 could be an option for WP3.0
			if ( $wp->query_vars['callback'] == 'ngg-ajax') {
				require_once (dirname (__FILE__) . '/xml/ajax.php');
				exit();
			}

		}

		function required_version() {

			global $wp_version;

			// Check for WP version installation
			$wp_ok  =  version_compare($wp_version, $this->minimum_WP, '>=');

			if ( ($wp_ok == FALSE) ) {
				add_action(
					'admin_notices',
					create_function(
						'',
						'global $ngg; printf (\'<div id="message" class="error"><p><strong>\' . __(\'Sorry, NextGEN Gallery works only under WordPress %s or higher\', "nggallery" ) . \'</strong></p></div>\', $ngg->minimum_WP );'
					)
				);
				return false;
			}

			return true;

		}

		function check_memory_limit() {

			// get the real memory limit before some increase it
			$this->memory_limit = ini_get('memory_limit');

			// PHP docs : Note that to have no memory limit, set this directive to -1.
			if ($this->memory_limit == -1 ) return true;

			// Yes, we reached Gigabyte limits, so check if it's a megabyte limit
			if (strtolower( substr($this->memory_limit, -1) ) == 'm') {

				$this->memory_limit = (int) substr( $this->memory_limit, 0, -1);

				//This works only with enough memory, 16MB is silly, wordpress requires already 16MB :-)
				if ( ($this->memory_limit != 0) && ($this->memory_limit < 16 ) ) {
					add_action(
						'admin_notices',
						create_function(
							'',
							'echo \'<div id="message" class="error"><p><strong>' . __('Sorry, NextCellent Gallery works only with a Memory Limit of 16 MB or higher', 'nggallery') . '</strong></p></div>\';'
						)
					);
					return false;
				}
			}

			return true;

		}


		function define_tables() {
			global $wpdb;

			// add database pointer
			$wpdb->nggpictures					= $wpdb->prefix . 'ngg_pictures';
			$wpdb->nggallery					= $wpdb->prefix . 'ngg_gallery';
			$wpdb->nggalbum						= $wpdb->prefix . 'ngg_album';

		}

		function register_taxonomy() {
			global $wp_rewrite;

			// Register the NextGEN taxonomy
			$args = array(
					'label' => __('Picture tag', 'nggallery'),
					'template' => __('Picture tag: %2$l.', 'nggallery'),
					'helps' => __('Separate picture tags with commas.', 'nggallery'),
					'sort' => true,
					'args' => array('orderby' => 'term_order')
					);

			register_taxonomy( 'ngg_tag', 'nggallery', $args );
		}

		function define_constant() {

			global $wp_version;

			//TODO:SHOULD BE REMOVED LATER
			define('NGGVERSION', $this->version);
			// Minimum required database version
			define('NGG_DBVERSION', $this->dbversion);

			// required for Windows & XAMPP
			define('WINABSPATH', str_replace("\\", "/", ABSPATH) );

			// define URL
			define('NGGFOLDER', basename( dirname(__FILE__) ) );

			define('NGGALLERY_ABSPATH', trailingslashit( str_replace("\\","/", WP_PLUGIN_DIR . '/' . NGGFOLDER ) ) );
			define('NGGALLERY_URLPATH', trailingslashit( plugins_url( NGGFOLDER ) ) );

			// look for imagerotator
			define('NGGALLERY_IREXIST', !empty( $this->options['irURL'] ));

			// get value for safe mode
			if ( (gettype( ini_get('safe_mode') ) == 'string') ) {
				// if sever did in in a other way
				if ( ini_get('safe_mode') == 'off' ) define('SAFE_MODE', FALSE);
				else define( 'SAFE_MODE', ini_get('safe_mode') );
			} else
			define( 'SAFE_MODE', ini_get('safe_mode') );

			if ( version_compare($wp_version, '3.2.999', '>') )
				define('IS_WP_3_3', TRUE);

		}

		function load_dependencies() {

			// Load global libraries												// average memory usage (in bytes)
			require_once (dirname (__FILE__) . '/lib/core.php');					//  94.840
			require_once (dirname (__FILE__) . '/lib/ngg-db.php');					// 132.400
			require_once (dirname (__FILE__) . '/lib/image.php');					//  59.424
			require_once (dirname (__FILE__) . '/lib/tags.php');				    // 117.136
			require_once (dirname (__FILE__) . '/lib/post-thumbnail.php');			//  n.a.
			require_once (dirname (__FILE__) . '/widgets/widgets.php');				// 298.792
			require_once (dirname (__FILE__) . '/lib/multisite.php');
			require_once (dirname (__FILE__) . '/lib/sitemap.php');

			// Load frontend libraries
			require_once (dirname (__FILE__) . '/lib/navigation.php');		        // 242.016
			require_once (dirname (__FILE__) . '/nggfunctions.php');		        // n.a.
			require_once (dirname (__FILE__) . '/lib/shortcodes.php'); 		        // 92.664

			//Just needed if you access remote to WordPress
			if ( defined('XMLRPC_REQUEST') )
				require_once (dirname (__FILE__) . '/lib/xmlrpc.php');

			// We didn't need all stuff during a AJAX operation
			if ( defined('DOING_AJAX') )
				require_once (dirname (__FILE__) . '/admin/ajax.php');
			else {
				require_once (dirname (__FILE__) . '/lib/meta.php');				// 131.856
				require_once (dirname (__FILE__) . '/lib/media-rss.php');			//  82.768
				require_once (dirname (__FILE__) . '/lib/rewrite.php');				//  71.936
				include_once (dirname (__FILE__) . '/admin/tinymce/tinymce.php'); 	//  22.408

				// Load backend libraries
				if ( is_admin() ) {
					require_once (dirname (__FILE__) . '/admin/admin.php');
					require_once (dirname (__FILE__) . '/admin/media-upload.php');
					$this->nggAdminPanel = new nggAdminPanel();
				}
			}
		}

		function load_textdomain() {

			load_plugin_textdomain('nggallery', false, NGGFOLDER . '/lang');

		}

		function load_scripts() {

			// if you don't want that NGG load the scripts, add this constant
			if ( defined('NGG_SKIP_LOAD_SCRIPTS') )
				return;

			//	activate Thickbox
			if ($this->options['thumbEffect'] == 'thickbox') {
				wp_enqueue_script( 'thickbox' );
				// Load the thickbox images after all other scripts
				add_action( 'wp_footer', array(&$this, 'load_thickbox_images'), 11 );

			}

			// activate jquery.lightbox
			if ($this->options['thumbEffect'] == 'lightbox') {
				wp_enqueue_script('jquery');
			}

			// activate modified Shutter reloaded if not use the Shutter plugin
			if ( ($this->options['thumbEffect'] == "shutter") && !function_exists('srel_makeshutter') ) {
				wp_register_script('shutter', NGGALLERY_URLPATH .'shutter/shutter-reloaded.js', false ,'1.3.3');
				wp_localize_script('shutter', 'shutterSettings', array(
							'msgLoading' => __('L O A D I N G', 'nggallery'),
							'msgClose' => __('Click to Close', 'nggallery'),
							'imageCount' => '1'
				) );
				wp_enqueue_script( 'shutter' );
			}

			// required for the slideshow
			if ( NGGALLERY_IREXIST == true && $this->options['enableIR'] == '1' && nggGallery::detect_mobile_phone() === false )
				wp_enqueue_script('swfobject');
			else {
				wp_register_script('jquery-cycle', NGGALLERY_URLPATH .'js/jquery.cycle.all.min.js', array('jquery'), '2.9995');
				wp_enqueue_script('ngg-slideshow', NGGALLERY_URLPATH .'js/ngg.slideshow.min.js', array('jquery-cycle'), '1.06');

			}

			// Load AJAX navigation script, works only with shutter script as we need to add the listener
			if ( $this->options['galAjaxNav'] ) {
				if ( ($this->options['thumbEffect'] == "shutter") || function_exists('srel_makeshutter') ) {
					wp_enqueue_script ( 'ngg_script', NGGALLERY_URLPATH . 'js/ngg.js', array('jquery'), '2.1');
					wp_localize_script( 'ngg_script', 'ngg_ajax', array('path'		=> NGGALLERY_URLPATH,
																		'callback'  => trailingslashit( home_url() ) . 'index.php?callback=ngg-ajax',
																		'loading'	=> __('loading', 'nggallery'),
					) );
				}
			}

			// If activated, add PicLens/Cooliris javascript to footer
			if ( $this->options['usePicLens'] )
				nggMediaRss::add_piclens_javascript();

		}

		function load_thickbox_images() {
			// WP core reference relative to the images. Bad idea
			echo "\n" . '<script type="text/javascript">tb_pathToImage = "' . site_url() . '/wp-includes/js/thickbox/loadingAnimation.gif";tb_closeImage = "' . site_url() . '/wp-includes/js/thickbox/tb-close.png";</script>'. "\n";
		}

		function load_styles() {

			// check first the theme folder for a nggallery.css
			if ( nggGallery::get_theme_css_file() )
				wp_enqueue_style('NextGEN', nggGallery::get_theme_css_file() , false, '1.0.0', 'screen');
			else if ($this->options['activateCSS'])
				wp_enqueue_style('NextGEN', NGGALLERY_URLPATH . 'css/' . $this->options['CSSfile'], false, '1.0.0', 'screen');

			//	activate Thickbox
			if ($this->options['thumbEffect'] == 'thickbox')
				wp_enqueue_style( 'thickbox');

			// activate modified Shutter reloaded if not use the Shutter plugin
			if ( ($this->options['thumbEffect'] == 'shutter') && !function_exists('srel_makeshutter') )
				wp_enqueue_style('shutter', NGGALLERY_URLPATH .'shutter/shutter-reloaded.css', false, '1.3.4', 'screen');

		}

		function load_options() {
			// Load the options
			$this->options = get_option('ngg_options');
		}

		// Add rewrite rules
		function start_rewrite_module() {
			global $nggRewrite;

			if ( class_exists('nggRewrite') )
				$nggRewrite = new nggRewrite();
		}

		// THX to Shiba for the code
		// See: http://shibashake.com/wordpress-theme/write-a-plugin-for-wordpress-multi-site
		function multisite_new_blog($blog_id, $user_id, $domain, $path, $site_id, $meta ) {
			global $wpdb;

			include_once (dirname (__FILE__) . '/admin/install.php');

			if (is_plugin_active_for_network( $this->plugin_name )) {
				$current_blog = $wpdb->blogid;
				switch_to_blog($blog_id);
				nggallery_install();
				switch_to_blog($current_blog);
			}
		}

		/**
		 * Removes all transients created by NextGEN. Called during activation
		 * and deactivation routines
		 */
		static function remove_transients()
		{
			global $wpdb, $_wp_using_ext_object_cache;

			// Fetch all transients
			$query = "
				SELECT option_name FROM {$wpdb->options}
				WHERE option_name LIKE '%ngg_request%'
			";
			$transient_names = $wpdb->get_col($query);;

			// Delete all transients in the database
			$query = "
				DELETE FROM {$wpdb->options}
				WHERE option_name LIKE '%ngg_request%'
			";
			$wpdb->query($query);

			// If using an external caching mechanism, delete the cached items
			if ($_wp_using_ext_object_cache) {
				foreach ($transient_names as $transient) {
					wp_cache_delete($transient, 'transient');
					wp_cache_delete(substr($transient, 11), 'transient');
				}
			}
		}

		function activate() {
			global $wpdb;
			//Starting from version 1.8.0 it's works only with PHP5.2
			if (version_compare(PHP_VERSION, '5.2.0', '<')) {
					deactivate_plugins($this->plugin_name); // Deactivate ourself
					wp_die("Sorry, but you can't run this plugin, it requires PHP 5.2 or higher.");
					return;
			}

			// Clean up transients
			self::remove_transients();

			include_once (dirname (__FILE__) . '/admin/install.php');

			if (is_multisite()) {
				$network=isset($_SERVER['SCRIPT_NAME'])?$_SERVER['SCRIPT_NAME']:"";
				$activate=isset($_GET['action'])?$_GET['action']:"";
				$isNetwork=($network=='/wp-admin/network/plugins.php')?true:false;
				$isActivation=($activate=='deactivate')?false:true;

				if ($isNetwork and $isActivation){
					$old_blog = $wpdb->blogid;
					$blogids = $wpdb->get_col($wpdb->prepare("SELECT blog_id FROM $wpdb->blogs", NULL));
					foreach ($blogids as $blog_id) {
						switch_to_blog($blog_id);
						nggallery_install();
					}
					switch_to_blog($old_blog);
					return;
				}
			}

			// check for tables
			nggallery_install();
			// remove the update message
			delete_option( 'ngg_update_exists' );

		}

		function deactivate() {

			// remove & reset the init check option
			delete_option( 'ngg_init_check' );
			delete_option( 'ngg_update_exists' );

			// Clean up transients
			self::remove_transients();
		}

		function uninstall() {
			// Clean up transients
			self::remove_transients();

			include_once (dirname (__FILE__) . '/admin/install.php');
			nggallery_uninstall();
		}

		function disable_upgrade($option){

			// PHP5.2 is required for NGG V1.4.0
			if ( version_compare($option->response[ $this->plugin_name ]->new_version, '1.4.0', '>=') )
				return $option;

			if( isset($option->response[ $this->plugin_name ]) ){
				//Clear it''s download link
				$option->response[ $this->plugin_name ]->package = '';

				//Add a notice message
				if ($this->add_PHP5_notice == false){
					add_action( "in_plugin_update_message-$this->plugin_name", create_function('', 'echo \'<br /><span style="color:red">Please update to PHP5.2 as soon as possible, the plugin is not tested under PHP4 anymore</span>\';') );
					$this->add_PHP5_notice = true;
				}
			}
			return $option;
		}

		// Add links to Plugins page
		function add_plugin_links($links, $file) {

			if ( $file == $this->plugin_name ) {
				$plugin_name = plugin_basename(NGGALLERY_ABSPATH);
				$links[] = "<a href='admin.php?page={$plugin_name}'>" . __('Overview', 'nggallery') . '</a>';
				$links[] = '<a href="http://wordpress.org/support/plugin/nextcellent-gallery-nextgen-legacy">' . __('Get help', 'nggallery') . '</a>';
				//$links[] = '<a href="">' . __('Contribute', 'nggallery') . '</a>';
			}
			return $links;
		}

		// Check for the header / footer, parts taken from Matt Martz (http://sivel.net/)
		function test_head_footer_init() {

			// If test-head query var exists hook into wp_head
			if ( isset( $_GET['test-head'] ) )
				add_action( 'wp_head', create_function('', 'echo \'<!--wp_head-->\';'), 99999 );

			// If test-footer query var exists hook into wp_footer
			if ( isset( $_GET['test-footer'] ) )
				add_action( 'wp_footer', create_function('', 'echo \'<!--wp_footer-->\';'), 99999 );
		}

		/**
		* Handles upload requests
		*/
		function handle_upload_request()
		{
			if (isset($_GET['nggupload'])) {
				require_once(implode(DIRECTORY_SEPARATOR, array(
					NGGALLERY_ABSPATH,
					'admin',
					'upload.php'
				)));
				throw new E_Clean_Exit();
			}
		}

		/**
		* Handles clean exits gracefully. Re-raises anything else
		* @param Exception $ex
		*/
		function exception_handler($ex)
		{
			if (get_class($ex) != 'E_Clean_Exit') throw $ex;
		}
	}

	// Let's start the holy plugin
	global $ngg;
	$ngg = new nggLoader();
}




/**
 * Checks if there is a NextGEN version running. If so, it deactivates itself
 * TODO: Has to be improved. error msg needs to be translated.
 */
class check_nextgen {

    static function nextgen_activated() {

        if (!function_exists('get_plugin_data')) {
            include_once( ABSPATH . 'wp-admin/includes/plugin.php' );
        }

        $nextcellent_plugin= plugin_basename(__FILE__);

        $plugin_list = get_plugins();

        //Loop over all the active plugins
        foreach ($plugin_list as $plugin_file=>$plugin_data) {
            //If we found nextcellent, skip it
            if ($plugin_file==$nextcellent_plugin) continue;
            //If the plugin is deactivate it ignore it.
            if (!is_plugin_active($plugin_file)) continue;
            if (strpos($plugin_file,'nggallery.php')!==FALSE) {
                $version = $plugin_data['Version'];
                //Check if effectively could be nextgen
                $is_nextgen= (strpos(strtolower($plugin_data['Name']),'nextgen') !==FALSE);
                if ($is_nextgen) { //is it?
                    //Yes, display msg on admin console
                    add_action(
                        'admin_notices',
                        create_function(
                            '',
                            'echo \'<div id="message" class="error"><p><strong>' . __('Sorry, NextCellent Gallery is deactivated: NextGEN version ' . $version . ' was detected. Deactivate it before running NextCellent!', 'nggallery') . '</strong></p></div>\';'

                         //sprintf( __('Client error when retrieving file:(%d) - %s'  , 'statcomm'),$code,$errMsg)
                        )
                    );
                    //Deactivate this plugin
                    deactivate_plugins($nextcellent_plugin);
                    return true;
               }
            }
        }
        return false;
    }
}
?>